parselglossy/validate.py

```diff
@@ -0,0 +1,195 @@
+    """
+    input_sections = [x for x in input_dict if type(input_dict[x]).__name__ == 'dict']
+    input_keywords = list(filter(lambda x: x not in input_sections, input_dict.keys()))
+
+    if 'sections' in template_dict:
+        template_sections = [x['section'] for x in template_dict['sections']]
+    else:
+        template_sections = []
+
+    if 'keywords' in template_dict:
+        template_keywords = [x['keyword'] for x in template_dict['keywords']]
+    else:
+        template_keywords = []
+
+    # stop if we find a keyword without documentation
+    keywords_no_doc = [x['keyword'] for x in template_dict['keywords']
+                       if 'documentation' not in x or x['documentation'].strip() == '']
+    if len(keywords_no_doc) > 0:
+        raise TemplateError('keyword(s) without any documentation: {}'.format(keywords_no_doc))
+
+    # make sure that we did not receive keywords
+    # or sections which we did not expect
+    difference = set(input_keywords).difference(set(template_keywords))
+    if difference != set():
+        raise InputError('found unexpected keyword(s): {}'.format(difference))
+    difference = set(input_sections).difference(set(template_sections))
+    if difference != set():
+        raise InputError('found unexpected section(s): {}'.format(difference))
+
+    # check that keywords without a default are set in the input
+    keywords_no_default = [x['keyword'] for x in template_dict['keywords'] if 'default' not in x]
+    difference = set(keywords_no_default).difference(set(input_keywords))
+    if difference != set():
+        raise InputError('the following keyword(s) must be set: {}'.format(difference))
+
+    # for each keyword verify the type of the value
+    for keyword in input_keywords:
+        template_keyword = list(filter(lambda x: x['keyword'] == keyword, template_dict['keywords']))[0]
+        _type = template_keyword['type']
+        if not type_matches(input_dict[keyword], _type):
+            raise InputError("incorrect type for keyword: '{0}', expected '{1}' type".format(keyword, _type))
+
+    # fill missing input keywords with template defaults
+    for keyword in set(template_keywords).difference(set(input_keywords)):
+        template_keyword = list(filter(lambda x: x['keyword'] == keyword, template_dict['keywords']))[0]
+        _default = template_keyword['default']
+        input_dict[keyword] = _default
+
+    # if this node contains sections, we descend into these and verify these in turn
+    for section in template_sections:
+        input_section = input_dict[section]
+        template_section = list(filter(lambda x: x['section'] == section, template_dict['sections']))[0]
+        input_dict[section] = validate_node(input_section, template_section)
+
+    return input_dict
+
+
+def check_predicates_node(input_dict: JsonDict,
+                          input_dict_node: JsonDict,
+                          template_dict_node: JsonDict) -> None:
```

Comment:
This function raises as soon as the first predicate fails, am I correct? As a final user I would prefer to know everything that's wrong with my input file, as the back and forth is quite unpleasant. Follow up and discussion in #24